Lib/glob.py

```diff
@@ -45,82 +42,36 @@ def iglob(pathname, *, root_dir=None, dir_fd=None, recursive=False,
-            if _lexists(_join(root_dir, pathname), dir_fd):
-                yield pathname
-        else:
-            # Patterns ending with a slash should match only directories
-            if _isdir(_join(root_dir, dirname), dir_fd):
-                yield pathname
-        return
-    if not dirname:
-        if recursive and _isrecursive(basename):
-            yield from _glob2(root_dir, basename, dir_fd, dironly,
-                             include_hidden=include_hidden)
-        else:
-            yield from _glob1(root_dir, basename, dir_fd, dironly,
-                              include_hidden=include_hidden)
-        return
-    # `os.path.split()` returns the argument itself as a dirname if it is a
-    # drive or UNC path.  Prevent an infinite recursion if a drive or UNC path
-    # contains magic characters (i.e. r'\\?\C:').
-    if dirname != pathname and has_magic(dirname):
-        dirs = _iglob(dirname, root_dir, dir_fd, recursive, True,
-                      include_hidden=include_hidden)
-    else:
-        dirs = [dirname]
-    if has_magic(basename):
-        if recursive and _isrecursive(basename):
-            glob_in_dir = _glob2
-        else:
-            glob_in_dir = _glob1
+    pathname = os.fspath(pathname)
+    if isinstance(pathname, bytes):
+        pathname = os.fsdecode(pathname)
+        if root_dir is not None:
+            root_dir = os.fsdecode(root_dir)
+        for path in _iglob(pathname, root_dir, dir_fd, recursive, include_hidden):
+            yield os.fsencode(path)
     else:
-        glob_in_dir = _glob0
-    for dirname in dirs:
-        for name in glob_in_dir(_join(root_dir, dirname), basename, dir_fd, dironly,
-                               include_hidden=include_hidden):
-            yield os.path.join(dirname, name)
-
-# These 2 helper functions non-recursively glob inside a literal directory.
-# They return a list of basenames.  _glob1 accepts a pattern while _glob0
-# takes a literal basename (so it only has to check for its existence).
-
-def _glob1(dirname, pattern, dir_fd, dironly, include_hidden=False):
-    names = _listdir(dirname, dir_fd, dironly)
-    if not (include_hidden or _ishidden(pattern)):
-        names = (x for x in names if not _ishidden(x))
-    return fnmatch.filter(names, pattern)
-
-def _glob0(dirname, basename, dir_fd, dironly, include_hidden=False):
-    if basename:
-        if _lexists(_join(dirname, basename), dir_fd):
-            return [basename]
+        yield from _iglob(pathname, root_dir, dir_fd, recursive, include_hidden)
+
+def _iglob(pathname, root_dir, dir_fd, recursive, include_hidden):
+    anchor, parts = _split_pathname(pathname)
+    globber = _StringGlobber(recursive=recursive, include_hidden=include_hidden)
+    select = globber.selector(parts)
+    if anchor:
+        # Non-relative pattern. The anchor is guaranteed to exist unless it
+        # has a Windows drive component.
+        exists = not os.path.splitdrive(anchor)[0]
```

Comment:
`splitroot()` is preferred since it's a builtin function, and `splitdrive()` just calls it and returns `drive, root + tail`. The `splitdrive()` call in `escape()` should also be replaced with `splitroot()`.

```suggestion
        exists = not os.path.splitroot(anchor)[0]
```

Reply:
The situation is reversed on POSIX - `splitroot()` is almost a no-op:

https://github.com/python/cpython/blob/34f5ae69fe9ab0f5b23311d5c396d0cbb5902913/Lib/posixpath.py#L131-L135

Reply:
`splitroot()` is a builtin function on all platforms. Even if `splitroot()` does more work on POSIX, it's made up for by being a builtin. I don't think it's not worth the extra cost of calling `splitdrive()` on Windows.

Linux:

```text
$ python -m timeit -s "import os" "os.path.splitroot('/spam')[0]"
5000000 loops, best of 5: 63.7 nsec per loop
$ python -m timeit -s "import os" "os.path.splitdrive('/spam')[0]"
5000000 loops, best of 5: 72.6 nsec per loop

$ python -m timeit -s "import os" "os.path.splitroot('//spam')[0]"
5000000 loops, best of 5: 76.5 nsec per loop
$ python -m timeit -s "import os" "os.path.splitdrive('//spam')[0]"
5000000 loops, best of 5: 73.5 nsec per loop
```

Windows: 

```text
> python -m timeit -s "import os" "os.path.splitroot('/spam')[0]"
2000000 loops, best of 5: 88 nsec per loop
> python -m timeit -s "import os" "os.path.splitdrive('/spam')[0]"
2000000 loops, best of 5: 174 nsec per loop

> python -m timeit -s "import os" "os.path.splitroot('//spam')[0]"
2000000 loops, best of 5: 105 nsec per loop
> python -m timeit -s "import os" "os.path.splitdrive('//spam')[0]"
2000000 loops, best of 5: 157 nsec per loop
```